pkcs11sec/pkcs11_security.go

```diff
@@ -0,0 +1,642 @@
+	prefix, ok := hashPrefixes[opts.HashFunc()]
+	if !ok {
+		return nil, errors.New("unknown hash function")
+	}
+	mechanism := pkcs11.NewMechanism(pkcs11.CKM_RSA_PKCS, nil)
+	input := append(prefix, msg...)
+
+	output, err := k.PrivateKey.Sign(*mechanism, input)
+	if err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+type Config struct {
+	// CAFile is the file where the trusted CA cert resides
+	CAFile string
+
+	// CertCacheDir is the dir where cached certs reside
+	CertCacheDir string
+
+	// PrivilegedUsers is a list of regular expressions that identity privileged users
+	PrivilegedUsers []string
+
+	// AllowList is a list of regular expressions that identity valid users to allow in
+	AllowList []string
+
+	// DisableTLSVerify disables TLS verify in HTTP clients etc
+	DisableTLSVerify bool
+
+	// AlwaysOverwriteCache supports always overwriting the local filesystem cache
+	AlwaysOverwriteCache bool
+
+	// PKCS11DriverFile points to the dynamic library file to use (usually a .so file)
+	PKCS11DriverFile string
+
+	// PKCS11Slot specifies which slot of the pkcs11 device to use
+	PKCS11Slot uint
+}
+
+func New(opts ...Option) (*Pkcs11Security, error) {
+	p := &Pkcs11Security{}
+
+	for _, opt := range opts {
+		err := opt(p)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if p.conf == nil {
+		return nil, errors.New("configuration not given")
+	}
+
+	if p.log == nil {
+		return nil, errors.New("logger not given")
+	}
+
+	if p.conf.PKCS11DriverFile == "" {
+		return nil, errors.New("pkcs11: PKCS11DriverFile option is required")
+	}
+
+	if p.pin != nil {
+		if err := p.loginToToken(); err != nil {
+			return nil, errors.Wrap(err, "failed to login to token in New()")
+		}
+	}
+
+	return p, p.reinit()
+}
+
+func promptForPin() (*string, error) {
+	byteStr, err := gopass.GetPasswdPrompt("PIN: ", false, os.Stdin, os.Stdout)
+	if err != nil {
+		return nil, err
+	}
+	str := string(byteStr)
+	return &str, nil
+}
+
+func (p *Pkcs11Security) loginToToken() error {
+
+	var err error
+	if p.pin == nil {
+		p.pin, err = promptForPin()
+		if err != nil {
+			return err
+		}
+	}
+
+	p.log.Debugf("Attempting to open PKCS11 driver file %s", p.conf.PKCS11DriverFile)
+
+	module, err := p11.OpenModule(p.conf.PKCS11DriverFile)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open PKCS11 driver file %s", p.conf.PKCS11DriverFile)
+	}
+
+	p.log.Debug("Attempting to fetch PKCS11 driver slots")
+
+	slots, err := module.Slots()
+	if err != nil {
+		return errors.Wrap(err, "failed to fetch PKCS11 driver slots")
+	}
+
+	var slot *p11.Slot
+	found := false
+	for _, aSlot := range slots {
+		p.log.Debugf("Found slot %d", aSlot.ID())
+
+		if aSlot.ID() == p.conf.PKCS11Slot {
+			slot = &aSlot
+			found = true
+			break
+		}
+	}
+	if !found {
+		if len(slots) == 1 {
+			slot = &slots[0]
+		} else {
+			return fmt.Errorf("failed to find slot with label %d", p.conf.PKCS11Slot)
```

Comment:
you use `errors.New()` mostly here for making new errors, so keep using that so its consistent :)